商品購入機能

FjiiNamiki opened the pull request:
# What
商品購入のためのコントローラーやビューファイルの編集

# Why
商品購入機能実装のため

- 正しく情報を記入すると、商品の購入ができる動画
https://gyazo.com/1686bd78e572d70bf0933f029abb4356
- 入力に問題のある状態では商品の購入はできず、エラーメッセージが出力される動画
https://gyazo.com/e21ac777389bc91d27a3def3a10bf39d
- ログアウト状態のユーザーが、URLを直接入力して商品購入ページに遷移しようとすると、商品の販売状況に関わらずログインページに遷移する動画
https://gyazo.com/92a7d5c685328133befc220c248a2985
- ログイン状態の出品者が、URLを直接入力して自身の出品した商品購入ページに遷移しようとすると、トップページに遷移する動画
https://gyazo.com/567c16f0fd22794b4f9904dd3482af3b
- ログイン状態のユーザーが、URLを直接入力して売却済み商品の商品購入ページへ遷移しようとすると、トップページに遷移する動画
https://gyazo.com/d59498efbd8188b8b0506efb7ed9bd12
- 売却済みの商品は、画像上に『sold out』の文字が表示される動画（商品一覧機能実装時に未実装であった場合）
https://gyazo.com/d91423c73602bb6933dcfcd10bf3ffd8
- 売却済みの商品は、画像上に『sold out』の文字が表示される動画（商品詳細機能実装時に未実装であった場合）
https://gyazo.com/af2b0ec19e462d2c6bf3caff58923417
- ログイン状態の出品者でも、売却済みの商品に対しては「編集・削除ボタン」が表示されない動画（商品詳細機能実装時に未実装であった場合）
https://gyazo.com/276aba2b655ae920dd7c25199b2b0681
- ログイン状態の出品者以外のユーザーでも、売却済みの商品に対しては「購入画面に進む」ボタンが表示されない動画（商品詳細機能実装時に未実装であった場合）
https://gyazo.com/9077c0e0aa46ce4b113bd66ec97c6aa4
- ログイン状態の出品者であっても、URLを直接入力して自身の売却済み商品の商品情報編集ページへ遷移しようとすると、トップページに遷移する動画（商品情報編集機能実装時に未実装であった場合）
https://gyazo.com/ad0bb621e332a703bd9288cc06e8cd93
- テスト結果の画像
https://gyazo.com/52b05e1aaf436cba810caca49b02707d

FjiiNamiki committed:
フォームオブジェクトとコントローラーを作成

FjiiNamiki committed:
データベースの作成

FjiiNamiki committed:
購入機能に必要なモデルやコントローラーを作成

FjiiNamiki committed:
購入のためのフォームを実装

FjiiNamiki committed:
フォームの情報を保存できるようにする

FjiiNamiki committed:
エラーが表示されるようにする

FjiiNamiki committed:
テストコードを実装

FjiiNamiki committed:
クレジット機能搭載

FjiiNamiki committed:
ルボコップ

FjiiNamiki committed:
提出前最終の修正

ushi10 commented:
レビュー依頼ありがとうございます！確認させていただきました。

挙動確認の動画ですが、以下のものが正確に確認ができなかったので改めて再度撮影をお願いできますでしょうか？
```
・ログイン状態の出品者が、URLを直接入力して自身の出品した商品購入ページに遷移しようとすると、トップページに遷移する動画
```
自身の出品した商品か否かが動画内では確認ができませんでしたので、こちらがわかるように撮影をお願いいたします。

動画を撮影後、改めてURLを貼り付けた状態で再度レビュー依頼フォームにてご提出をお願いいたします。

FjiiNamiki commented:
すみません。貼り直しました。
-ログイン状態の出品者が、URLを直接入力して自身の出品した商品購入ページに遷移しようとすると、トップページに遷移する動画
https://gyazo.com/6f56211a1e86626d7da9a6dcedd72c99

yutomentors commented on app/views/items/index.html.erb:
```diff
@@ -136,9 +136,11 @@
           <%= image_tag(item.image, class: "item-img") %>
 
           <%# 商品が売れていればsold outを表示しましょう %>
```
```
<%# 商品が売れていればsold outを表示しましょう %>
```

こちらのコメントアウトですが、不要であれば削除しましょう。

理由：
企業様に提出する可能性のあるアプリに学習用の実装案内や使用しない記述が残っている状態が好ましくないためです。

yutomentors commented on app/views/items/index.html.erb:
```diff
@@ -136,9 +136,11 @@
             <span>Sold Out!!</span>
           </div>
+          <% end %> 
           <%# //商品が売れていればsold outを表示しましょう %>
```
```
<%# 商品が売れていればsold outを表示しましょう %>
```

こちらのコメントアウトですが、不要であれば削除しましょう。

理由：
企業様に提出する可能性のあるアプリに学習用の実装案内や使用しない記述が残っている状態が好ましくないためです。

yutomentors commented on app/views/items/show.html.erb:
```diff
@@ -9,9 +9,11 @@
     <div class="item-img-content">
       <%= image_tag(@item.image, class:"item-box-img") %>
       <%# 商品が売れている場合は、sold outを表示しましょう %>
```
```
<%# 商品が売れていればsold outを表示しましょう %>
```

こちらのコメントアウトですが、不要であれば削除しましょう。

理由：
企業様に提出する可能性のあるアプリに学習用の実装案内や使用しない記述が残っている状態が好ましくないためです。

yutomentors commented on app/views/items/show.html.erb:
```diff
@@ -9,9 +9,11 @@
+          <span>Sold Out!!</span>
+        </div>
+      <% end %>
       <%# //商品が売れている場合は、sold outを表示しましょう %>
```
```
<%# 商品が売れていればsold outを表示しましょう %>
```

こちらのコメントアウトですが、不要であれば削除しましょう。

理由：
企業様に提出する可能性のあるアプリに学習用の実装案内や使用しない記述が残っている状態が好ましくないためです。

yutomentors commented on app/views/items/show.html.erb:
```diff
@@ -23,15 +25,15 @@
         <p class="or-text">or</p>
         <%= link_to "削除", item_path(@item.id), method: :delete, class:"item-destroy" %>
 
         <%# 商品が売れていない場合はこちらを表示しましょう %>
```
```
<%# 商品が売れていない場合はこちらを表示しましょう %>
```

こちらのコメントアウトですが、不要であれば削除しましょう。

理由：
企業様に提出する可能性のあるアプリに学習用の実装案内や使用しない記述が残っている状態が好ましくないためです。

yutomentors commented on app/views/items/show.html.erb:
```diff
@@ -23,15 +25,15 @@
+        <%= link_to "購入画面に進む", item_orders_path(@item.id), method: :get, class:"item-red-btn"%>
       <% end %>
     <% end %>
     <%# //商品が売れていない場合はこちらを表示しましょう %>
```
```
<%# 商品が売れていない場合はこちらを表示しましょう %>
```

こちらのコメントアウトですが、不要であれば削除しましょう。

理由：
企業様に提出する可能性のあるアプリに学習用の実装案内や使用しない記述が残っている状態が好ましくないためです。

yutomentors commented on spec/models/order_destination_spec.rb:
```diff
@@ -0,0 +1,72 @@
+        @order_destination.valid?
+        expect(@order_destination.errors.full_messages).to include "Phonenumber can't be blank"
+      end
+      it 'phonenumberが11桁以上の数字では登録できない' do
```
電話番号のテストコードですが以下のように細分化しましょう。

```
電話番号が9桁以下だと登録できないこと
電話番号が12桁以上だと登録できないこと
```


理由
日本には10桁の電話番号（固定電話）が存在するため、
11桁でしか保存できない設定にしていると、
利用者にとって不都合が生じる可能性があります。

yutomentors commented on spec/models/order_destination_spec.rb:
```diff
@@ -0,0 +1,72 @@
+        @order_destination.phonenumber = 'ああ'
+        @order_destination.valid?
+        expect(@order_destination.errors.full_messages).to include "Phonenumber is invalid"
+      end
```
以下のテストを追加しましょう！
```
userが紐づいていなければ登録できない
itemが紐づいていなければ登録できない
```

yutomentors reviewed (commented):
レビュー依頼ありがとうございます！
コメントしましたので、ご確認をお願いいたします！

修正頂いた後、コミットプッシュを確認いただき、再度レビュー依頼のご提出をお願い致します！
修正後の最新のテスト実行結果の画像も添付いただきますようお願いいたします。

FjiiNamiki committed:
商品購入機能修正1

# What
1. 不要なコメントアウトの削除
1. 電話番号のテストコードを細分化
1. アソシエーションのテストコードを追加

 # Why
1. 企業に提出際に学習用のコメントアウトが存在していることは良くないため。
1. テストコードの精度を上げるため。
1. テストコードの精度を上げるため。

FjiiNamiki commented:
レビューありがとうございます。修正しました。
# What
1. 不要なコメントアウトの削除
1. 電話番号のテストコードを細分化
1. アソシエーションのテストコードを追加

# Why
1. 企業に提出際に学習用のコメントアウトが存在していることは良くないため。
1. テストコードの精度を上げるため。
1. テストコードの精度を上げるため。

yutomentors reviewed (commented):
修正とレビュー依頼ありがとうございます。

問題なく実装できていることが確認できましたので、
LGTMとさせていただきます！

こちらのブランチは忘れずにMergeを行っていただき、
ローカルのMasterにPullしておきましょう。

ここまでの実装、本当にお疲れ様でした。
引き続き、応援しております。
